File: _07_Counting_sort.cpp
#include<iostream>
using namespace std;

void printArray(int array[], int n);
//counting sort:
void countingSort(int array[], int n, int k){
    int countArray[k+1] = {0};
    int helperArray[n];
    //calculate occurrences :
    for (int i = 0; i < n; i++)
    {
        countArray[array[i]]++;
    }
    //calculate positions :
    for (int i = 1; i <= k; i++)
    {
        countArray[i] = countArray[i] + countArray[i-1];
    }
    //building helper array :
    for (int i = n-1; i >= 0; i--)
    {
        helperArray[--countArray[array[i]]] = array[i];
    }
    //coping to real array : 
    for (int i = 0; i < n; i++)
    {
        array[i] = helperArray[i];
    }
}

int main(){
    int n;
    cout << "Enter the size of the array : \n";
    cin >> n;
    int array[n+1];
    int k = 0;
    cout << "Enter the array : \n";
    for(int i = 0; i < n; i++)
    {
        cin >> array[i];
        if(array[i]>k)
            k = array[i];
    }

    countingSort(array, n, k);

    cout << "\nAfter Counting Sorting :\n";
    printArray(array,n);
}

//printing : 
void printArray(int array[], int n){
    for (int i = 0; i < n; i++)
    {
        cout << array[i]<< " ";
    }cout << endl;
}

File: _07_Counting_sort.java
import java.util.Scanner;

public class _07_Counting_sort {
    public static void main(String[] args) {
        int[] array = input();
        int n = array.length;
        int k = max(array,n);

        countingSort(array,n,k);
        System.out.println("\nAfter Counting Sorting : ");
        printArray(array,n);
    }

//Counting sort:
    public static void countingSort(int[] array, int n, int k){
        int[] countingArray = new int[k+1];
        int[] helperArray = new int[n];
        //compute countingArray:
        for(int i = 0; i < n; i++){
            countingArray[array[i]]++;
        }
        //compute the locations:
        for(int i = 1; i <= k; i++){
            countingArray[i] = countingArray[i]+countingArray[i-1];
        }
        //building helperArray:
        for(int i = n-1; i >= 0; i--){
            helperArray[--countingArray[array[i]]] = array[i];
        }
        //coping to original array:
        for(int i = 0; i < n; i++){
            array[i] = helperArray[i];
        }
    }

    //findig k :
    public static int max(int[]array, int n){
        int mx = 0;
        for(int i = 0; i < array.length; i++){
            if(array[i]>mx){
                mx = array[i];
            }
        }
        return mx;
    }

    //Taking input : 
    public static int[] input(){
        int n;
        System.out.println("Enter the size of the array : ");
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        int[] array = new int[n];
        System.out.println("Enter the array : ");
        for(int i = 0; i < n; i++){
            array[i] = scanner.nextInt();
        }
        scanner.close();
        return array;
    }

    //Print array : 
    public static void printArray(int[] array, int n){
        for(int i = 0; i < n; i++){
            System.out.print(array[i]+" ");
        }System.out.println();
    }
}
